### PR TITLE
[5.7] improve heuristics for @main support

### PR DESCRIPTION
**Pull request by tomerd**
5.7 cherrypick of #5687 and #5704

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci smoke test

**Review by neonichu — APPROVED** [review]

**Comment by abertelrud** [comment]
These heuristics make sense to me.  FWIW, I cleaned up some of my earlier requests about a `-parse-as-executable` flag at https://github.com/apple/swift/issues/60277 so that build systems can declare the expected intent and don't need to be in the business of trying to conditionally defeat the compiler's heuristics.

**Comment by abertelrud** [comment]
But for now this looks like a definite improvement.

File: Sources/Build/BuildPlan.swift [hunk 1]
```diff
@@ -626,26 +626,57 @@ public final class SwiftTargetBuildDescription {
...
-    /// See https://bugs.swift.org/browse/SR-14488 for discussion about improvements so that SwiftPM can
-    /// convey the intent to build an executable module to the compiler regardless of the number of files
-    /// in the module or their names.
+    /// of the source code
     var needsToBeParsedAsLibrary: Bool {
-        switch target.type {
+        switch self.target.type {
         case .library, .test:
             return true
         case .executable:
```
Comment (abertelrud) [inline Sources/Build/BuildPlan.swift:634]: Does this also need to apply to snippets BTW?  They're a lot like executables, AFAIK?

**Review by abertelrud — APPROVED** [review]

File: Sources/Build/BuildPlan.swift (hunk 1 above)
Comment (zkiraly) [reply Sources/Build/BuildPlan.swift:634]: I think so. Please see the last code example here: https://github.com/zkiraly/swift-argument-parser/blob/main/Sources/ArgumentParser/Documentation.docc/Articles/GettingStarted.md

With Xcode 14 Beta 4, if this example is implemented as a snippet, I get errors.

Comment (neonichu) [reply Sources/Build/BuildPlan.swift:634]: None of this will apply to Xcode BTW, since we're not using the Build module there.

**Comment by bitjammer** [comment]
@tomerd Could you possibly pull https://github.com/apple/swift-package-manager/pull/5704 into this?

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
> @tomerd Could you possibly pull https://github.com/apple/swift-package-manager/pull/5704 into this?

@bitjammer done

**Pull request merged**